go-controller/cmd/ovn-kube-cni-overlay/ovn-kube-cni-overlay.go

```diff
@@ -0,0 +1,255 @@
+		}
+		err = netlink.LinkSetHardwareAddr(link, hwAddr)
+		if err != nil {
+			return fmt.Errorf("failed to add mac address %s to %s: %v", macAddress, contIfaceName, err)
+		}
+
+		addr, err := netlink.ParseAddr(ipAddress)
+		if err != nil {
+			return err
+		}
+		err = netlink.AddrAdd(link, addr)
+		if err != nil {
+			return fmt.Errorf("failed to add IP addr %s to %s: %v", ipAddress, contIfaceName, err)
+		}
+
+		gw := net.ParseIP(gatewayIP)
+		if gw == nil {
+			return fmt.Errorf("parse ip of gateway failed")
+		}
+		err = ip.AddRoute(nil, gw, link)
+		if err != nil {
+			return err
+		}
+
+		hostIfaceName = hostVeth.Name
+
+		return nil
+	})
+	if err != nil {
+		return "", err
+	}
+
+	// rename the host end of veth pair
+	newHostIfaceName := containerID[:15]
+	if err := renameLink(hostIfaceName, newHostIfaceName); err != nil {
+		return "", fmt.Errorf("failed to rename %s to %s: %v", hostIfaceName, newHostIfaceName, err)
+	}
+
+	return newHostIfaceName, nil
+}
+
+func argString2Map(args string) (map[string]string, error) {
+	argsMap := make(map[string]string)
+
+	pairs := strings.Split(args, ";")
+	for _, pair := range pairs {
+		kv := strings.Split(pair, "=")
+		if len(kv) != 2 {
+			return nil, fmt.Errorf("ARGS: invalid pair %q", pair)
+		}
+		keyString := kv[0]
+		valueString := kv[1]
+		argsMap[keyString] = valueString
+	}
+
+	return argsMap, nil
+}
+
+// TODO: use a better way to connect to api server
+func getAnnotationOnPod(server, namespace, pod string) (map[string]string, error) {
+	url := fmt.Sprintf("%s/api/v1/namespaces/%s/pods/%s", server, namespace, pod)
+
+	res, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	var apiPod kapi.Pod
+	err = json.NewDecoder(res.Body).Decode(&apiPod)
+	if err != nil {
+		return nil, err
+	}
+
+	return apiPod.ObjectMeta.Annotations, nil
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	argsMap, err := argString2Map(args.Args)
+	if err != nil {
+		return err
+	}
+
+	namespace := argsMap["K8S_POD_NAMESPACE"]
+	podName := argsMap["K8S_POD_NAME"]
+	containerID := argsMap["K8S_POD_INFRA_CONTAINER_ID"]
+
+	if namespace == "" || podName == "" || containerID == "" {
+		return fmt.Errorf("required CNI variable missing")
+	}
+
+	ovsArgs := []string{
+		"--if-exists", "get", "Open_vSwitch",
+		".", "external_ids:k8s-api-server",
+	}
+	out, err := exec.Command("ovs-vsctl", ovsArgs...).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to get K8S_API_SERVER")
+	}
+	k8sAPIServer := strings.Trim(strings.TrimSpace(string(out)), "\"")
+	if !strings.HasPrefix(k8sAPIServer, "http") {
+		k8sAPIServer = fmt.Sprintf("http://%s", k8sAPIServer)
+	}
+
+	// Get the IP address and MAC address from the API server.
+	// Wait for a maximum of 3 seconds with a retry every 0.1 second.
+	var annotation map[string]string
+	for cnt := 0; cnt < 30; cnt++ {
+		annotation, err = getAnnotationOnPod(k8sAPIServer, namespace, podName)
```

Comment:
Even though we get the annotation from the pod, it could still not have the OVN annotation. So the retry mechanism should keep trying till we get the OVN annotation.